### ISSUE TITLE
DuckDB binary crashed at duckdb::ExpressionIterator::EnumerateChildren

### ISSUE THREAD

**Issue by fuboat** [duckdb/duckdb#8185]
### What happens?

The DuckDB binary (/usr/local/bin/duckdb) crashed in `duckdb::ExpressionIterator::EnumerateChildren`.
The binary compiled with RelWithDebInfo without ASAN reports "''duckdb' terminated by signal SIGSEGV (Address boundary error)".
The binary compiled with RelWithDebInfo and ASAN reports it as stack-overflow.

### To Reproduce
```sql
CREATE TABLE v0 ( v1 INTEGER , v2 INTEGER) ; 
SELECT * FROM v0 WHERE v2 <= 2 AND v2 <= v1 AND v1 < v2 ORDER BY v1 DESC ;
```
### OS:

ubuntu 20.04 x64

### DuckDB Version:

v0.8.2-dev1293 f2dc7d8bce

### DuckDB Client:

Binary (/usr/local/bin/duckdb) v0.8.2-dev1293 f2dc7d8bce

### Full Name:

Jingzhou Fu

### Affiliation:

Wingtecher Lab of Tsinghua University and Shuimuyulin ltd.

### Have you tried this on the latest `master` branch?

- [X] I agree

### Have you tried the steps to reproduce? Do they include all relevant data and configuration? Does the issue you report still appear there?

- [X] I agree

**Comment by fuboat** [comment]
@Mytherin @Tishj Perhaps this bug report was inadvertently missed?

I supplement the ASAN report in the latest commit version (v0.8.2-dev2044 762396f87b) here:

```
AddressSanitizer:DEADLYSIGNAL
=================================================================
==4790==ERROR: AddressSanitizer: stack-overflow on address 0x7fffebb70fb8 (pc 0x556e95b7a4c6 bp 0x7fffebb713c0 sp 0x7fffebb70fb0 T0)
    #0 0x556e95b7a4c5 in duckdb::ExpressionIterator::EnumerateChildren(duckdb::Expression&, std::function<void (duckdb::unique_ptr<duckdb::Expression, std::default_delete<duckdb::Expression>, true>&)> const&) /root/duckdb_master/src/planner/expression_iterator.cpp:23
    #1 0x556e95b7a19b in duckdb::ExpressionIterator::EnumerateChildren(duckdb::Expression const&, std::function<void (duckdb::Expression const&)> const&) /root/duckdb_master/src/planner/expression_iterator.cpp:15
    #2 0x556e95b85bf1 in duckdb::Expression::Hash() const /root/duckdb_master/src/planner/expression.cpp:94
    #3 0x556e9582156c in duckdb::BoundColumnRefExpression::Hash() const /root/duckdb_master/src/planner/expression/bound_columnref_expression.cpp:24
    #4 0x556e93115d53 in duckdb::ExpressionHashFunction<duckdb::Expression>::operator()(std::reference_wrapper<duckdb::Expression> const&) const /root/duckdb_master/src/include/duckdb/parser/expression_map.hpp:23
    #5 0x556e957aaace in std::__detail::_Hash_code_base<std::reference_wrapper<duckdb::Expression>, std::pair<std::reference_wrapper<duckdb::Expression> const, duckdb::unique_ptr<duckdb::Expression, std::default_delete<duckdb::Expression>, true> >, std::__detail::_Select1st, duckdb::ExpressionHashFunction<duckdb::Expression>, std::__detail::_Mod_range_hashing, std::__detail::_Default_ranged_hash, true>::_M_hash_code(std::reference_wrapper<duckdb::Expression> const&) const /usr/include/c++/9/bits/hashtable_policy.h:1384
AddressSanitizer:DEADLYSIGNAL
AddressSanitizer: nested bug in the same thread, aborting.
```

**Comment by Mytherin** [comment]
It is on our list of fuzzer issues, we will get to it before the next release

**Comment by Mytherin** [comment]
Thanks for the report! I've pushed a fix in #9440. The issue here was an infinite recursion in an optimizer caused by the inverted conditions (`v2 <= v1 AND v1 < v2`).
